src/index.ts

```diff
@@ -312,18 +325,19 @@ export function ruleFactory<
       rule,
       ignoreMissingKeys = false,
     }: {
-      stepRow: number;
-      input: TInput;
-      rule: string | string[] | Rule;
-      ignoreMissingKeys?: boolean;
+      stepRow: number
+      input: TInput
+      rule: string | string[] | Rule
+      ignoreMissingKeys?: boolean
     }): RuleResult {
+      // checking only the first rule seems unsafe
```

Comment:
It's about time this index zero check gets more robust. 😅 

What do you think about a composable utility, roughly:

```ts
const validateRules = {
  isString: rule => typeof rule === 'string' && rule.length > 4,
  isStringOrStringList: rule => Array.isArray(rule) ? rule.every(validateRules.isString) : validateRules.isString(rule),
  isLogicalRule: rule => rule != null && ('and' in rule || 'or' in rule),
  isStructuralRule: rule => rule != null && ('if' in rule || 'and' in rule || 'or' in rule || 'map' in rule || 'reduce' in rule || 'filter' in rule || 'return' in rule),
  // TODO: add more combined rules to express higher order structures, constraints, etc
}
```